sympy/solvers/solveset.py

```diff
@@ -68,18 +68,16 @@ def _masked(f, *atoms):
     >>> f
     cos(cos(x) + 1)
     """
-    sym = numbered_symbols('a', cls=Dummy)
+    sym = numbered_symbols('a', cls=Dummy, real=True)
     mask = []
     for a in ordered(f.atoms(*atoms)):
         for i in mask:
             a = a.replace(*i)
         mask.append((a, next(sym)))
-    mask = list(reversed(mask))
     for i, (o, n) in enumerate(mask):
         f = f.replace(o, n)
         mask[i] = (n, o)
-    if mask and f == mask[0][1]:
-        f = mask[0][0]
+    mask = list(reversed(mask))
```

Comment:
Slightly improved the logic for `_masked`, it seemed that equations like these: `2*Abs(x + Abs(x + Max(3, x))`, were not actually maked properly, it returned `2*Abs(_a0 + x) - 2` rather than `_a1`. This was due to the fact that masking was taking place with outside arguments (Abs) first then inside arguments, it should have started masking off from inner arguments.

Reply:
Nice catch.